protocol/src/protocol.typeHierarchy.proposed.md

```diff
@@ -0,0 +1,222 @@
+#### Type Hierarchy
+
+The type hierarchy language feature provides type hierarchy information for languages supporting inheritance and nominal typing.
+
+_Client Capabilities_:
+
+```ts
+/**
+ * Client capabilities specific to the type hierarchy feature.
+ */
+export interface TypeHierarchyCapabilities {
+
+  /**
+   * The text document client capabilities.
+   */
+  textDocument?: {
+
+    /**
+     * Capabilities specific to the `textDocument/typeHierarchy`.
+     */
+    typeHierarchy?: {
+
+      /**
+       * Whether implementation supports dynamic registration. If this is set to `true`
+       * the client supports the new `(TextDocumentRegistrationOptions & StaticRegistrationOptions)`
+       * return value for the corresponding server capability as well.
+       */
+      dynamicRegistration?: boolean;
+
+    }
+
+  }
+
+}
+```
+
+_Server Capabilities_:
+
+Capability whether the language server capability can provider type hierarchy information to the client.
+
+```ts
+/**
+ * Type hierarchy language server capability.
+ */
+export interface TypeHierarchyServerCapabilities {
+
+  /**
+   * Server capability for calculating super- and subtype hierarchies.
+   */
+  typeHierarchyProvider?: boolean | (TextDocumentRegistrationOptions & StaticRegistrationOptions);
+
+}
+```
+
+_Requests_:
+
+The `Type Hierarchy Request` is sent from the client to the server to retrieve a `TypeHierarchyItem` at a given cursor position. If computing full type hierarchy items is expensive, servers can additionally provide a handler for the type hierarchy item resolve request (`typeHierarchy/resolve`). This request would also allow to specify if the item should be resolved and whether sub- or supertypes are to be resolved. If no item can be retrieved for a given text document position, returns with `null`.
```

Comment:
This seems pretty complicated - is there evidence that a) lazy resolving is needed and b) eager resolving is *also* needed?